### PR TITLE
Add `LightningCLI(run=False|True)`

### LINKED ISSUE #8615: LightningCLI shouldn't run fit by default
### To Reproduce

Currently, the LightingCLI runs fit by default which is an anti-pattern for the Lightning API and seems like magic.

Current implementation.
```py
    cli = LightningCLI(LiftModel, MnistDataModule)
    cli.trainer.test(datamodule=cli.datamodule)
    cli.trainer.save_checkpoint("mnist_cnn.pt")
```

Expected implementation.
```py
    cli = LightningCLI(LiftModel, MnistDataModule)
    cli.trainer.fit(cli.model, datamodule=cli.datamodule)
    cli.trainer.test(datamodule=cli.datamodule)
    cli.trainer.save_checkpoint("mnist_cnn.pt")
```

### Environment

 - PyTorch Lightning Version (e.g., 1.3.0):
 - PyTorch Version (e.g., 1.8)
 - Python version:
 - OS (e.g., Linux):
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - How you installed PyTorch (`conda`, `pip`, source):
 - If compiling from source, the output of `torch.__config__.show()`:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #8615, #8251
Part of #7508 

After discussing with @mauvilsa, we came to the conclusion that the solution of having a `__init__(run: bool)` flag to disable `fit` is better than doing it by not passing a subcommand:

```python
# what this PR implements
# LightningCLI(..., run=False)
$ python script.py ...

# vs

# what was the previous idea
# LightningCLI(...)
$ python script.py ...
```

This is because there is no case where the user would go from doing `python script.py fit <some_args>` to `python script.py <some_args>` without code changes because to do so, the user would need to write `cli.trainer.fit(cli.model)` anyways after the CLI instantiation.

With that in consideration, we require the user to explicitly set `run=False`.
This way, you have to really mess up to run `fit` twice accidentally.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: tests/utilities/test_cli.py [hunk 1]
```diff
@@ -687,3 +687,12 @@ def __init__(self, optim1: dict, optim2: dict, scheduler: dict):
     assert isinstance(cli.model.optim1, torch.optim.Adam)
     assert isinstance(cli.model.optim2, torch.optim.SGD)
     assert isinstance(cli.model.scheduler, torch.optim.lr_scheduler.ExponentialLR)
+
+
+@pytest.mark.parametrize("run", (False, True))
+def test_lightning_cli_disabled_run(run):
+    with mock.patch("sys.argv", ["any.py"]), mock.patch("pytorch_lightning.Trainer.fit") as fit_mock:
+        cli = LightningCLI(BoringModel, run=run)
+    fit_mock.assert_called() if run else fit_mock.assert_not_called()
```
Comment (awaelchli) [inline tests/utilities/test_cli.py:696]:
```suggestion
    fit_mock.call_count == run
```
as an alternative way to check it

File: docs/source/common/lightning_cli.rst [hunk 2]
```diff
@@ -260,17 +260,30 @@ file. Loading a defaults file :code:`my_cli_defaults.yaml` in the current workin
...
 
 To load a file in the user's home directory would be just changing to :code:`~/.my_cli_defaults.yaml`. Note that this
 setting is given through :code:`parser_kwargs`. More parameters are supported. For details see the `ArgumentParser API
 <https://jsonargparse.readthedocs.io/en/stable/#jsonargparse.core.ArgumentParser.__init__>`_ documentation.
 
 
+Instantiation only mode
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The CLI is designed to start fitting without minimal code changes. On class instantiation, the CLI will automatically
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:273]:
```suggestion
The CLI is designed to start fitting without code changes. On class instantiation, the CLI will automatically
```
or

```suggestion
The CLI is designed to start fitting with minimal code changes. On class instantiation, the CLI will automatically
```
?

File: docs/source/common/lightning_cli.rst [hunk 3]
```diff
@@ -260,17 +260,30 @@ file. Loading a defaults file :code:`my_cli_defaults.yaml` in the current workin
 
 .. testcode::
 
-    cli = LightningCLI(
-        MyModel,
-        MyDataModule,
-        parser_kwargs={"default_config_files": ["my_cli_defaults.yaml"]},
-    )
+    cli = LightningCLI(MyModel, MyDataModule, parser_kwargs={"default_config_files": ["my_cli_defaults.yaml"]})
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:263]: I didn't know it supports multiple config files. That looks very useful!

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]
looks the same as mine #8251 which you blocked :P

**Review by mauvilsa — APPROVED** [review]
Since there will be multiple ways of using LightningCLI there is one thing that could be done to prevent mistakes. When `run==True` after running, the Trainer instance could be modified overriding the `fit`, `test`, etc. methods such that it shows an error instead of running.

**Comment by carmocca** [comment]
> after running, the Trainer instance could be modified overriding the fit, test, etc. methods such that it shows an error instead of running.

This would limit those who want to fit with `run=True` and still call `test` manually later (or fit again!)

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/utilities/cli.py [hunk 4]
```diff
@@ -213,6 +213,7 @@ def __init__(
         parser_kwargs: Dict[str, Any] = None,
         subclass_mode_model: bool = False,
         subclass_mode_data: bool = False,
+        run: bool = True,
```
Comment (ananthsub) [inline pytorch_lightning/utilities/cli.py:216]: super nit: for other CLI tools i've seen `dry_run` be the term used. we could use it here too

Comment (carmocca) [reply pytorch_lightning/utilities/cli.py:216]: AFAIK, `dry_run` is sort of a "test it tworks" flag before doing the actual run

However, this `run` flag isn't for those who want to test their parsing, but for those who don't want to automatically start fitting their model.

Also notice that this flag is not exposed via command line. So you couldn't do

```python
# see if it parses
python script.py --dry_run True

# actual fit
python script.py
```

You'd need to change the file's flag.

File: docs/source/common/lightning_cli.rst [hunk 5]
```diff
@@ -260,17 +260,30 @@ file. Loading a defaults file :code:`my_cli_defaults.yaml` in the current workin
...
 
+Instantiation only mode
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The CLI is designed to start fitting with minimal code changes. On class instantiation, the CLI will automatically
+call ``trainer.fit(...)`` internally so you don't have to do it. To avoid this, you can set the following argument:
+
+.. testcode::
+
+    cli = LightningCLI(MyModel, run=False)  # True by default
```
Comment (tchaton) [inline docs/source/common/lightning_cli.rst:278]: @carmocca I believe this should be False by default.

**Review by tchaton — CHANGES_REQUESTED** [review]
Hey @carmocca. 

Can you share why run is set to True by default ? 
This seems counter-intuitive as you could argue that test should also be run too.
Furthermore, it doesn't fully resolve #8615 as it will run fit by default.
IMO, I would prefer user to opt-in on a running fit to avoid confusion.

**Comment by carmocca** [comment]
> Can you share why run is set to True by default ?

Because the entire design is focused on getting up-and-running as easily as possible.

Setting it to `False` would require a full restructure of the docs.

> Furthermore, it doesn't fully resolve #8615 as it will run fit by default.

We can remove that issue from the "fixes" list. This PR support the choice

> This seems counter-intuitive as you could argue that test should also be run too.

The CLI docs only mention that it is designed to run `fit` by default so that is documented.

> you could argue that test should also be run too.

This is entirely fair, running `fit` only was just the initial design. #7508 will add support for `test`

> I would prefer user to opt-in on a running fit to avoid confusion.

Again, #7508 will make it opt-in as providing a subcommand will be a requirement with `run=True` (default)

**Review by tchaton — APPROVED** [review]
